nomad/job_endpoint.go

```diff
@@ -939,6 +942,14 @@ func (j *Job) BatchDeregister(args *structs.JobBatchDeregisterRequest, reply *st
 		return err
 	}
 
+	// This RPC endpoint is only called from the Nomad server leader when
+	// performing garbage collection, therefore we can get away with a simple
+	// management check to ensure the caller can trigger this functionality as
+	// the leader token is always a management token.
+	if !aclObj.IsManagement() {
+		return structs.ErrPermissionDenied
+	}
```

Comment:
The comment here isn't _quite_ right, as the GC eval can be processed on a follower using the leader's token. But because we never expect a human operator to use this RPC (it's not exposed in the HTTP API), we can narrow this further to a server op check, similar to the plan applier, etc.:

```suggestion
	// This RPC endpoint is only called during garbage collection
	if !aclObj.AllowServerOp() {
		return structs.ErrPermissionDenied
	}
```

Reply:
I did consider changing the ACL check to the server only method, but I had one reservation about how this would work during an upgrade.

In a 3 server situation:
- one follower is upgraded to the new code
- the upgraded follower picks up the core eval to run the GC 
- the upgraded follower submits the GC write to the leader without adding an ACL token to the request
- the leader running the old code attempts to perform the namespace capability check, which fails

The ACL token addition to the RPC request could be kept which I believe would avoid the situation described above, but I am skeptical of adding this due to the following comment within the `AuthenticateServerOnly` function:

> This should never be used for RPCs that serve HTTP
// endpoints or accept ACL tokens to avoid confused deputy attacks by making a
// request to a follower that's forwarded.

Keen to know you thoughts here and whether the comment would impact upgrades where a non-upgrade follower submits a request to an upgraded leader?

Reply:
:facepalm: You're absolutely right. I was thinking that we could use the `AllowServerOp` with the `Authenticate` method that's already here rather than `AuthenticateServerOnly`, but that's not the intent at all. Nevermind!